kernel/vmbytegen.ml

```diff
@@ -778,7 +778,17 @@ let rec compile_lam env cenv lam sz cont =
     (* Hack: brutally pierce through the abstraction of PArray *)
     let dummy = KerName.make (ModPath.MPfile DirPath.empty) (Names.Label.of_id @@ Id.of_string "dummy") in
     let dummy = (MutInd.make1 dummy, 0) in
-    let ar = Lmakeblock (dummy, 0, args) in (* build the ocaml array *)
+    let ar, cont = match Vmlambda.as_value 0 args with
+    | None ->
+      (* build the ocaml array *)
+      Lmakeblock (dummy, 0, args), cont
+    | Some v ->
+      (* dump the blob as is, but copy the resulting parray afterwards so that
+         it does not get modified under our feet *)
+      let lbl = Label.create () in
+      (* dummy label, the array will never be an accumulator *)
+      Lval v, Klabel lbl :: Kcamlprim (CAML_Arraycopy, lbl) :: cont
```

Comment:
This cannot be correct. `Parray.copy` performs the copy of a persistent array, not of an OCaml array.

Reply:
I checked the code, it actually copies the inner array.

Reply:
Maybe you misunderstood this piece of code: `cont` is not the immediate continuation, it's what the program does after the parray-like value is built. In normal mode we build a parray from scratch out of a fresh block, but when the argument is a blob we reuse it as the inner block of the parray, then we copy the whole nested structure at the end. It so happens that `PArray.copy` behaves as a fresh copy of the inner array on arguments that need no rerooting, which is the case here.

Reply:
Sorry, you are right, I thought you were calling `Parray.copy` with argument `Lval v`.

I suggest changing a bit the comment. Indeed, I thought "it" was referring to the array, while it is referring to "the blob", which makes a lot more sense. And to be even clearer: "so that the blob is left untouched by modifications to the resulting parray".

Reply:
Done.